### PR TITLE
feat(codelens): Add codelens to run and debug tasks from cake files

### PR DESCRIPTION
**Pull request by jrgcubano**
Implement #80 

Codelens to run tasks from cake files:

- Codelens:
  - Can be turned on/off from vscode cake codelens settings
- Run task
   - Settings can be configured from vscode cake codelens settings (verbosity, etc)    
- Debug task
   - First verify if `debug extensions` are already installed and call the debug command to auto install the dependencies if needed.
   - Settings can be configured from vscode cake codelens settings (framework, etc) 

vscode-cake-debug-codelens

vscode-cake-codelens

package.json settings:

```
...
 "cake.codeLens": {
          "type": "object",
          "description": "The Cake codelens configuration",
          "default": {
            "showCodeLens": true,
            "scriptsIncludePattern": "**/*.cake",
            "taskRegularExpression": "Task\\s*?\\(\\s*?\"(.*?)\"\\s*?\\)",
            "debugTask": {
              "verbosity": "normal",
              "debugType": "coreclr",
              "request": "launch",
              "program": "${workspaceRoot}/tools/Cake.CoreCLR/Cake.dll",
              "cwd": "${workspaceRoot}",
              "stopAtEntry": true,
              "console": "internalConsole",
              "logging": {
                "exceptions": false,
                "moduleLoad": false,
                "programOutput": false,
                "engineLogging": false,
                "browserStdOut": false
              }
            },
            "runTask": {
              "verbosity": "normal"
            }
          },
...
```

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -123,6 +126,172 @@
...
+          },
+          "properties": {
+            "showCodeLens": {
+              "type": "boolean",
+              "description": "Show codelens for tasks on cake files",
+              "default": true
+            },
+            "scriptsIncludePattern": {
+              "type": "string",
+              "description": "Glob pattern to detect cake build scripts on current worksppace",
```
Comment (gep13) [inline package.json]: typo - worksppace -> workspace

File: package.json [hunk 2]
```diff
@@ -123,6 +126,172 @@
...
+              }
+            },
+            "runTask": {
+              "verbosity": "normal"
+            }
+          },
+          "properties": {
+            "showCodeLens": {
+              "type": "boolean",
+              "description": "Show codelens for tasks on cake files",
```
Comment (gep13) [inline package.json]: grammar - for tasks in Cake files

File: package.json (hunk 1 above)
Comment (gep13) [reply package.json]: grammar -> build scripts in current workspace

File: package.json [hunk 3]
```diff
@@ -123,6 +126,172 @@
...
+              "default": true
+            },
+            "scriptsIncludePattern": {
+              "type": "string",
+              "description": "Glob pattern to detect cake build scripts on current worksppace",
+              "default": "**/*.cake"
+            },
+            "taskRegularExpression": {
+              "type": "string",
+              "description": "Regular expression pattern to get tasks from cake script",
```
Comment (gep13) [inline package.json]: Perhaps uppercase C on Cake

File: package.json [hunk 4]
```diff
@@ -123,6 +126,172 @@
...
+                  "default": "coreclr"
+                },
+                "request": {
+                  "type": "string",
+                  "description": "The request type of the debug session",
+                  "default": "launch"
+                },
+                "program": {
+                  "type": "string",
+                  "description": "The program of the debug session",
```
Comment (gep13) [inline package.json]: grammar -> The executable for the debug session

File: package.json [hunk 5]
```diff
@@ -123,6 +126,172 @@
...
+                  "default": "${workspaceRoot}/tools/Cake.CoreCLR/Cake.dll"
+                },
+                "cwd": {
+                  "type": "string",
+                  "description": "Path to the working directory of the program being debugged. Default is the current workspace.",
```
Comment (gep13) [inline package.json]: Should this be the Cake script being debugged?

File: package.json [hunk 6]
```diff
@@ -123,6 +126,172 @@
...
+                      "default": false
+                    },
+                    "engineLogging": {
+                      "type": "boolean",
+                      "description": "Optional flag to determine whether diagnostic engine logs should be logged to the output window.",
+                      "default": false
+                    },
+                    "browserStdOut": {
+                      "type": "boolean",
+                      "description": "Optional flag to determine if stdout text from the launching the web browser should be logged to the output window.",
```
Comment (gep13) [inline package.json]: I think you have an extra "the" in here.

File: README.md [hunk 7]
```diff
@@ -59,15 +75,30 @@ There are a number of configuration options which allow you to control how the T
...
-Cake (C# Make) is a cross platform build automation system with a C# DSL to do things like compiling code, copy files/folders, running unit tests, compress files and build NuGet packages.
-
-## Learn more
-
-For more information about Cake, please see the [Cake website](https://cakebuild.net) or the Cake [source code repository](https://github.com/cake-build/cake).
+* `cake.taskRunner.taskRegularExpression`: a regular expression pattern which is used to identify Tasks within the `*.cake` files. Default value is `Task\\s*?\\(\\s*?\"(.*?)\"\\s*?\\)`.
+
+### Codelens
+
+The extension using codelens let you run and debug tasks individually on any `*.cake` file in your current workspace.
```
Comment (gep13) [inline README.md]: The extension uses codelens to allow you to run and debug tasks individually on any `*.cake` file in your current workspace.

File: README.md [hunk 8]
```diff
@@ -59,15 +75,30 @@ There are a number of configuration options which allow you to control how the T
...
+
+There are a number of configuration options which allow you to control the cake output verbosity and what host framework to use for the debug session, etc.
```
Comment (gep13) [inline README.md]: Capital C for Cake?

File: README.md [hunk 9]
```diff
@@ -59,15 +75,30 @@ There are a number of configuration options which allow you to control how the T
...
+* `cake.codeLens.scriptsIncludePattern`: a glob pattern which specifies how to detect `.cake` files in the current workspace. Default value is `**/*.cake`.
+* `cake.codeLens.taskRegularExpression`: a regular expression pattern which is used to identify Tasks within the `*.cake` files. Default value is `Task\\s*?\\(\\s*?\"(.*?)\"\\s*?\\)`.
+* `cake.codeLens.runTask.verbosity`: allows you to control cake `run task` verbosity (`diagnostic`, `minimal`, `normal`, `quiet` and `verbose`. Default value is `normal`.
+* `cake.codeLens.debugTask.verbosity`: allows you to control cake `debug task` verbosity (`diagnostic`, `minimal`, `normal`, `quiet` and `verbose`. Default value is `normal`.
+* `cake.codeLens.debugTask.debugType`: "framework type of the debug session (`mono`or `coreclr`). Default value is `coreclr`.
+* `cake.codeLens.debugTask.request`: request type of the debug session. Default value is `launch`.
+* `cake.codeLens.debugTask.program`: program of the debug session (`tools/Cake/Cake.exe` for `mono` or `tools/Cake.CoreCLR/Cake.dll` for `coreclr`). Default value is `${workspaceRoot}/tools/Cake.CoreCLR/Cake.dll`.
```
Comment (gep13) [inline README.md]: Perhaps executable is better here?

**Comment by jrgcubano** [comment]
@gep13 nice! done with corrections.

**Pull request merged**